pkg/standalone-utils/contracts/StableLPOracle.sol

```diff
@@ -0,0 +1,212 @@
+
+        tvl = 0;
+        for (uint256 i = 0; i < _totalTokens; i++) {
+            tvl += prices[i].toUint256().mulDown(marketPriceBalancesScaled18[i]);
+        }
+
+        return tvl;
+    }
+
+    /**
+     * @dev Computes a set of balances for a given price vector and a given invariant. The set of balances should meet
+     * two conditions:
+     * 1. The invariant of the computed balances should be equal to the given invariant.
+     * 2. The spot prices of the computed balances should be equal to the given price vector. In other words:
+     *    `spotPrice(i, j) == prices[i] / prices[j]`
+     */
+    function _computeMarketPriceBalances(
+        uint256 invariant,
+        int256[] memory prices
+    ) internal view returns (uint256[] memory balancesForPrices) {
+        // To compute the balances for a given price vector, we need to compute the gradient of the stable invariant.
+        // The stable invariant is:
+        //
+        // a \sum{xi} \prod{xi} +bD \prod{xi} - D^(n+1) = 0
+        //
+        // where `D` is the invariant, `n` is the number of tokens, `xi` is the balance of each token,
+        // `a = A*(n^2n)` and `b = a - n^n`.
+        //
+        // The gradient in terms of xj (the balance of the j-th token) is:
+        //
+        // a \prod{xi} + a \sum{xi} \prod_i!=j{xi} + bD \prod_i!=j{xi}
+        //
+        // We can make this gradient equal to k*pj, where pj is the price of the j-th token and k is a constant.
+        // Then, solving this system of equations for every pj, we will have an array of balances that respect the
+        // price vector.
+
+        (int256 a, int256 b) = _computeAAndBForPool(IStablePool(address(pool)));
+
+        // First, we need to compute the constant k that will multiply the prices.
+        int256 k = _computeK(a, b, prices);
+
+        int256 sumPriceDivision = 0;
+        for (uint256 i = 0; i < _totalTokens; i++) {
+            sumPriceDivision += _divDownInt(a, _mulDownInt(k, prices[i]) - a);
+        }
+
+        balancesForPrices = new uint256[](_totalTokens);
+        for (uint256 i = 0; i < _totalTokens; i++) {
+            balancesForPrices[i] = ((b * int256(invariant)) /
+                _mulDownInt(a - _mulDownInt(k, prices[i]), _ONE_INT - sumPriceDivision)).toUint256();
+        }
+    }
+
+    function _computeK(int256 a, int256 b, int256[] memory prices) internal view returns (int256 k) {
+        // k is computed by solving the equation:
+        //
+        // f(k) = T^(n+1)P - alpha = 0
+        //
+        // where `T=1+\sum{a/(a-k*pi)}`, `P = \prod{(a - k*pi)/a}` and `alpha=a*(b/a)^(n+1)`.
+        // Notice that P*T^n can be a very big number. To avoid math overflows, but keep the precision, we divide
+        // f(k) and f'(k) by P*T^n.
+
+        k = _findInitialGuessForK(a, b, prices);
+        for (uint256 i = 0; i < 255; i++) {
+            // dTdk and dPdk are the derivatives of T and P with respect to k.
```

Comment:
These names are a little "Curvy" and hard to follow.

Using the convention of using prime for derivatives; e.g.,  a' (1st derivative) and a'' (2nd derivative), which we don't have here:

// tPrimeOverK = first derivative of T(totalSomething) with respect to K
T = totalSomething (descriptive name); dTdk = tPrimeOverK; dPdk = pPrimeOverK

It's not easy, but slightly longer names would be less cryptic and easier to follow. Otherwise your eyes just glaze over. Descriptive names plus ASCII art on top of it would help a lot.